[feature] Filters v1

VyrCossont opened the pull request:
# Description

This pull request implements a client-side filter API compatible with the Mastodon filter v1 API.

It does not implement the Mastodon filter v2 API, although it does add the DB and cache plumbing to trivially implement that API later. It also doesn't implement the server-side filtering part of the v1 API: the `irreversible` filter option will return an error message if set to `true`.

Fixes #1472

## Checklist

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [x] I/we have made any necessary changes to documentation.
- [x] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

VyrCossont commented:
There are two cache-related TODOs: one is for the multi-invalidate cache method that doesn't exist yet, and one is for cache memory ratios, which I don't understand.

The Swagger tool GtS uses seems to have a lot of bugs and omissions, so I'm not 100% on the Swagger parts.

VyrCossont commented:
Using the CI for `golangci-lint run` because it's 100% broken on my dev box. Sorry!

VyrCossont commented on internal/gtsmodel/filter.go:
```diff
@@ -0,0 +1,71 @@
+	CreatedAt time.Time `bun:"type:timestamptz,nullzero,notnull,default:current_timestamp"`                  // when was item created
+	UpdatedAt time.Time `bun:"type:timestamptz,nullzero,notnull,default:current_timestamp"`                  // when was item last updated
+	AccountID string    `bun:"type:CHAR(26),notnull,nullzero"`                                               // ID of the local account that created the filter keyword.
+	FilterID  string    `bun:"type:CHAR(26),notnull,nullzero,unique:filter_keywords_filter_id_keyword_uniq"` // ID of the filter that this keyword belongs to.
```
After factoring out `FilterEntry` from `FilterKeyword` and `FilterStatus`, these uniqueness constraints won't work. Can we add them in the migration?

VyrCossont commented on internal/api/model/filterv1.go:
```diff
@@ -0,0 +1,100 @@
+
+// FilterCreateUpdateRequestV1 captures params for creating or updating a v1 filter.
+//
+// swagger:model filterCreateUpdateRequestV1
```
This should use https://goswagger.io/use/spec/params.html instead.

VyrCossont replied:
Swagger/OpenAPI 2 seems like it *really* wasn't built to handle the sloppy mess that is Rails-compatible "you can send this as form params or JSON" REST APIs; maybe OpenAPI 3 does better, but for now we're just going to do what the rest of the GtS codebase does: the currently prevailing approach is to use `swagger:ignore` on the objects representing the parsed versions of form parameters, and document the form parameters.

JJGadgets commented:
Please do let me know if I'm being a nuisance, but I really want this feature badly: is there any chance of this making it into 0.14?

Unfollowing accounts just because one of their views don't align or because they're reporting on news I'm not interested in, isn't a solution for me.

Also, more of a curiosity question, why only implement v1 (client-side) initially and not go straight to implementing v2 (server-side)?

VyrCossont commented:
Client-side filtering doesn't need any server work past filter storage, so it's low risk, and many Mastodon API clients already support it (including GtS favorites Semaphore and Feditext), so it's already useful.

There are some newer clients that only support server-side filtering, such as Phanpy, so I designed this patch's storage layer to make it easy to add in a followup.

VyrCossont commented:
@JJGadgets aside from the missing index, this should work fine on 0.14 if you want to apply it yourself. I've been running the 0.13 prototype on my own instance for weeks.

JJGadgets commented:
> @JJGadgets aside from the missing index, this should work fine on 0.14 if you want to apply it yourself. I've been running the 0.13 prototype on my own instance for weeks.

@VyrCossont should be safe to just build directly from your branch... right? (Those failed checks kept me on edge haha)

I'm already building my own distroless container images for GoToSocial and other apps anyway, so it shouldn't be too hard.

(I build my own images mainly for the fun of it, and because my environment is Kubernetes on Talos Linux which I use because of its minimal and secure distro design for running Kubernetes and nothing more on the host level, but to be clear, nothing's inherently wrong with the upstream image, distroless isn't for everyone.)

JJGadgets commented:
Also, do you know if v1 filters are compatible with Tusky and Elk?

I mainly use Mona (which isn't compatible with either kinds of filters), Tusky and Phanpy depending on which OS I'm on, but I'm okay with switching back from Phanpy to Elk if it means I can get filters.

VyrCossont commented:
@JJGadgets not sure about Tusky or Elk, sorry. When you find out, please let me know. And Mona doesn't have filter support at *all*? Yikes. Try Feditext.

VyrCossont commented:
Oops, that close was entirely accidental.

VyrCossont commented:
I take back my "this should work fine on 0.14"; there are some DB changes I need to rebase on top of first.

VyrCossont commented:
The unused code warnings are a false positive, possibly due to dominikh/go-tools#1440, and I've fixed the issue with GoLand custom folding region comments in go-critic/go-critic#1408. For now, I've added very specific exclusions to our linter config.

JJGadgets commented:
> @JJGadgets not sure about Tusky or Elk, sorry. When you find out, please let me know. And Mona doesn't have filter support at *all*? Yikes. Try Feditext.

Mona has filters but they don't talk to the instance at all, they're synced over iCloud instead and thus only work on Mona. Which I currently use because GtS doesn't support filters (before this PR), but that means my filtered timeline only shows on my Apple devices.

I'm in the Feditext beta, yeah. I generally prefer Mona though.

Will find out about Tusky and Elk and report back when I do.

tsmethurst commented:
> The unused code warnings are a false positive, possibly due to [dominikh/go-tools#1440](https://github.com/dominikh/go-tools/issues/1440), and I've fixed the issue with GoLand custom folding region comments in [go-critic/go-critic#1408](https://github.com/go-critic/go-critic/pull/1408). For now, I've added very specific exclusions to our linter config.
> 
> I have no idea why the docs build failed, but I haven't changed anything in the docs.

I'll check it out :)

tsmethurst commented on test/envparsing.sh:
```diff
@@ -31,6 +31,9 @@ EXPECT=$(cat << "EOF"
         "boost-of-ids-mem-ratio": 3,
         "emoji-category-mem-ratio": 0.1,
         "emoji-mem-ratio": 3,
+        "filter-keyword-mem-ratio": 1,
```
These ratios are probably a little higher than they need to be, I'd say 0.5 or something for each of them would probably be OK. Happy to defer to @NyaaaWhatsUpDoc though since she knows more about how these values end up being used.

VyrCossont replied:
Happy to defer as well.

VyrCossont replied:
Dropping these to 0.5 after discussion in Matrix.

tsmethurst commented on internal/api/client/filters/v1/filterdelete.go:
```diff
@@ -0,0 +1,92 @@
+//
+//	security:
+//	- OAuth2 Bearer:
+//		- write:lists
```
Think you mean write:filters here.

tsmethurst commented on internal/api/client/filters/v1/filterdelete.go:
```diff
@@ -0,0 +1,92 @@
+		return
+	}
+
+	id := c.Param(IDKey)
```
For parsing IDs and stuff like that, you can use the utility functions in apiutil like this:

```go
	id, errWithCode := apiutil.ParseID(c.Param(apiutil.IDKey))
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}
```

It doesn't really lead to fewer lines of code per se, but it does mean the errors we return are a bit more consistently worded.

Just note that `apiutil.IDKey` is "id" and not "ID".

tsmethurst commented on internal/api/client/filters/v1/filterput.go:
```diff
@@ -0,0 +1,160 @@
+//
+//	security:
+//	- OAuth2 Bearer:
+//		- read:filters
```
Should be "write" again.

tsmethurst commented on internal/api/model/filterv1.go:
```diff
@@ -0,0 +1,105 @@
+	// When the filter should no longer be applied. Not present if the filter does not expire.
+	//
+	// Example: 2024-02-01T02:57:49Z
+	ExpiresAt string `json:"expires_at,omitempty"`
```
I think this needs to a *string and be null if it doesn't expire, instead of omitted, see: https://docs.joinmastodon.org/entities/Filter/#expires_at

VyrCossont replied:
You're correct. Sample output from Mastodon:

```json
[
  {
    "id": "145781",
    "phrase": "test",
    "context": [
      "home",
      "public"
    ],
    "whole_word": true,
    "expires_at": null,
    "irreversible": false
  }
]
```

tsmethurst commented on internal/api/model/filterv1.go:
```diff
@@ -0,0 +1,105 @@
+type FilterV1 struct {
+	// The ID of the filter in the database.
+	//
+	// Required: true
```
Unless a filter is used as the body of a request somewhere, I don't think we need the `Required` tags on this model.

VyrCossont replied:
These are only used in responses so we can drop that.

tsmethurst commented on internal/db/filter.go:
```diff
@@ -0,0 +1,105 @@
+)
+
+type Filter interface {
+	//region Filter methods
```
What do these region comments do? Or are they just for clarity while reading?

VyrCossont replied:
They're semi-standard markers for custom code folding. So they're for clarity in reading, but also, in editors that are aware of region markers (I use IntelliJ/GoLand), you can do this:

![this file with two regions folded up to a single line with their name, and one unfolded and showing code](https://github.com/superseriousbusiness/gotosocial/assets/26195552/c474d65f-5dd2-4715-b193-229a1e866e24)

tsmethurst commented on internal/gtsmodel/filter.go:
```diff
@@ -0,0 +1,71 @@
+	UpdatedAt            time.Time        `bun:"type:timestamptz,nullzero,notnull,default:current_timestamp"` // when was item last updated
+	ExpiresAt            time.Time        `bun:"type:timestamptz,nullzero"`                                   // Time filter should expire. If null, should not expire.
+	AccountID            string           `bun:"type:CHAR(26),notnull,nullzero"`                              // ID of the local account that created the filter.
+	Title                string           `bun:",nullzero,notnull,unique:filter_title_uniq"`                  // The name of the filter.
```
If you're just making this one field unique on its own, without combining it with another field, you don't need to name the unique constraint; you can just tag it as `,unique` and let the db implementation name it for you.

tsmethurst commented on internal/gtsmodel/filter.go:
```diff
@@ -0,0 +1,71 @@
+type FilterKeyword struct {
+	FilterEntry
+	Keyword   string `bun:",nullzero,notnull"` // The keyword or phrase to filter against.
+	WholeWord bool   `bun:",notnull"`          // Should the filter consider word boundaries?
```
Probably good to set a default value in the db for this, eg:

``go
WholeWord *bool `bun:",nullzero,notnull,default:false"`
``

tsmethurst commented on internal/gtsmodel/filter.go:
```diff
@@ -0,0 +1,71 @@
+import "time"
+
+// Filter stores a filter created by a local account.
+type Filter struct {
```
We've had issues with booleans in bun before, so the sort of best practice we arrived at was to be very explicit about bools in tags. As such, I'd recommend making bools in these filter models into ptrs, and adding default values for each of them (false in every case I think?):

``
*bool `bun:",nullzero,notnull,default:false"`
``

VyrCossont replied:
That's super weird! I'll go make those changes.

tsmethurst commented on internal/db/bundb/filter.go:
```diff
@@ -0,0 +1,655 @@
+		NewSelect().
+		Model((*gtsmodel.Filter)(nil)).
+		Column("id").
+		Where("account_id = ?", accountID).
```
Generally speaking in all these functions instead of doing `Where("whatever = ?", whatever)`, it's better to do `Where("? = ?", bun.Ident("whatever"), whatever)`, so that it can be properly escaped by bun depending on which database implementation is being used.

VyrCossont replied:
Have we had a ton of identifier collisions with builtins, or identifiers that have to be quoted for other reasons? I'll still make the changes but I'm curious because I've never seen this pattern before for *known* IDs. (Arbitrary IDs, totally understandable.)

daenney replied:
I think it's nice because it avoids treating some identifiers are special or trusted. If you always do it this way, then you never have to think about it and you don't risk a situation in which you forget and boom SQL injection.

tsmethurst replied:
> Have we had a ton of identifier collisions with builtins, or identifiers that have to be quoted for other reasons?

Not collisions really, but we have had cases where generated SQL queries fail for one db driver and work for another, due to different escaping methods, which has lead to lots of confusion when testing. Can't remember any specific examples off the top of my head, but it's easier to just avoid potential issues entirely by hinting to bun that they'd idents.

tsmethurst commented on internal/db/bundb/filter.go:
```diff
@@ -0,0 +1,655 @@
+
+// entry wraps a pointer to a model extending [gtsmodel.FilterEntry]
+// and provides operations common across those models.
+type entry[Model any, Self any] interface {
```
By this I mean, I can see you're cooking up these generic types to avoid code repetition here, but it ends up being quite complex and difficult to reason about (for me, at least, I struggle to understand what's happening here). In my view it might be better to just put up with the repetition instead of trying to do these interface workarounds.

VyrCossont replied:
No problem. We should have adequate test coverage to de-generify it without breaking anything if that's the consensus. I personally find shorter code with less repetition easier to read and maintain, but this generic type has only two concrete instantiations so I'm fine doing it either way.

tsmethurst replied:
Personally I always lean towards writing more verbose but more explicit code, and I think this is sort of facilitated by Go as well, given how difficult the language tends to make it to do the clever generics stuff you can do in other languages (as you write in a comment here, you had to do a workaround to get this working). So my preference would be to de-generify it, but I'll wait to see if kim has any opinions on the matter. Thanks for your patience with our curmudgeonly, deliberate style :P

VyrCossont replied:
I've committed a de-generified version for comparison. If we decide we want to keep the generic version, we can revert a few specific commits, but now we can see both at once: commits up to f162d22 for generic and up to c69b420 for non-generic.

tsmethurst commented on internal/db/bundb/migrations/20240126064004_add_filters.go:
```diff
@@ -0,0 +1,105 @@
+				"filter_keywords": {
+					"filter_keywords_account_id_idx":         {false, []string{"account_id"}},
+					"filter_keywords_filter_id_idx":          {false, []string{"filter_id"}},
+					"filter_keywords_filter_id_keyword_uniq": {true, []string{"filter_id", "keyword"}},
```
For these unique indexes, it might just be better to set `unique` constraints on the appropriate fields of the models in `gtsmodel` and let the database create the constraints (and accompanying indexes) instead.

tsmethurst replied:
Mmm, right, I see what you mean. The common nested struct is a little neater, but I think I'd rather we be more explicit and just inline those models and use the DB tags. Then it's also clearer by just looking at the models which fields have uniqueness constraints on them, without having to look at the db structure or look through the migrations. Would you mind making the change?

tsmethurst commented:
Not sure why the docs build failed, but I'll kick it :)

VyrCossont commented on internal/validate/formvalidation.go:
```diff
@@ -44,6 +44,7 @@ const (
 	maximumProfileFieldLength     = 255
 	maximumProfileFields          = 6
 	maximumListTitleLength        = 200
+	maximumFilterKeywordLength    = 40
```
Actually I think we can remove this entirely.

Most implementations (both client-side filters and probably our future server-side filters) are going to weld all your keywords together as alternate branches of a compiled regex for fast matching, so the real limit is how much the regex engine can take, which is as much down to how many filters you have as how long they are.

tsmethurst commented:
This is looking almost ready to go :eyes: Just a few discussions left to resolve. Thanks for your continued hard work @VyrCossont !

NyaaaWhatsUpDoc commented on internal/api/client/filters/v1/validate.go:
```diff
@@ -0,0 +1,68 @@
+		case float64:
+			form.ExpiresIn = util.Ptr(int(e))
+
+		case string:
```
should we be expecting this as a string? this feels wrong. or is this mastodon normalized behaviour we're stuck with :p

VyrCossont replied:
This is the same set of types accepted by the poll expiration time mechanism, so I assume we've seen that happen at least once 😞

NyaaaWhatsUpDoc commented on internal/db/bundb/filter.go:
```diff
@@ -0,0 +1,345 @@
+	filters = slices.DeleteFunc(filters, func(filter *gtsmodel.Filter) bool {
+		if err := f.populateFilter(ctx, filter); err != nil {
+			// %w is allowed here.
+			//goland:noinspection GoPrintFunctions
```
is therre any way around needing to add this to every single `errs.Appendf()` with %w operator?

VyrCossont replied:
There may be, but I haven't found it yet.

nikclayton commented:
Hi - now this is merged is 0.15.0 the first release it will appear in? Or if you released a hypothetical 0.14.1 patch release would it appear there?

(Broadly, do you make patch branches from HEAD, or from the tag of the minor release you're patching?)

I'm asking so I can add it to https://github.com/pachli/pachli-android/blob/32a6d3465b19e90898e59cd21fa29f12941341d7/core/network/src/main/kotlin/app/pachli/core/network/Server.kt#L277

tsmethurst commented:
0.15.0

nikclayton commented:
Thanks @tsmethurst , https://github.com/pachli/pachli-android/pull/523 adds the support in Pachli. And thanks @VyrCossont for the implementation work (and keeping the Swagger definitions up to date, that's incredibly useful).
